### DISCUSSION TITLE
a temporary rollback to supported track elements if said element is not so.

### DISCUSSION THREAD

**Discussion by TheClaudeQC** [OpenRCT2/OpenRCT2#19086] [Feature requests]
so, you build a good hybrid/single rail and then notice you cannot save it because of a zero g-roll, medium loop or a large corkscrew?

well, here is a idea I have, I don't know how hard it could be. I will add few images of what incompatible tracks could transform to:

**Zero G-rolls** into barrel rolls "of their original direction" (the start/end of it is at the lowest unit of the zero g-roll (in absouloute units we can see in vanilla RCT2 as well (E. G. : if a track piece (here a zero g-roll) is at 0,75m on his lowest, it will go back to 1,5m to be still at allowed heights for the barrel roll))) : 
![Parc sans nom 2023-01-08 09-06-20](https://user-images.githubusercontent.com/83843734/211200994-a548e889-e6fb-4be3-a1bc-ed8ddc836a10.png)

**Large zero g-rolls** into a steep to gentle piece (or inverted order) and a barrel roll "same rule as Zero g-roll" in the start/second at (if the sequence is level to gentle, then gentle to steep, then large zero g-roll) 5 absoulute unit higher than the base :
![Parc sans nom 2023-01-08 09-06-29](https://user-images.githubusercontent.com/83843734/211201362-78a94af2-b8cd-442f-8ade-87658fe45aa9.png)

**Medium Loops** into a large loop "same rule again" :
![Parc sans nom 2023-01-08 09-06-48](https://user-images.githubusercontent.com/83843734/211201531-32512ebf-15ca-47b3-9a59-fadc52b56a40.png)

**Large Corkscrews** into standards ones "guess what? Same rule again!" :
![Parc sans nom 2023-01-08 09-06-37](https://user-images.githubusercontent.com/83843734/211201574-c864c0b1-e146-4f65-926d-4788a0cee198.png)

here is what they could look like when saving whit them :
![Parc sans nom 2023-01-08 09-09-30](https://user-images.githubusercontent.com/83843734/211201604-2935de90-03df-48c9-bec2-a87e6cba3693.png)
![Parc sans nom 2023-01-08 09-10-15](https://user-images.githubusercontent.com/83843734/211201608-b2bdc3dd-5034-4bda-95ef-9c0c9eace309.png)
colours are in order for the track (Primary, Secondary, Supports) ; (bright red, dark red, saturated red)

Whit an special warning like this ;
_**WARNING!!! some track elements aren't supported, they will roll back to supported ones when you will rebuild/view the design**_

**Comment by spacek531** [comment]
Does TD6 store the position data of elements? I do not believe it does since it can just dead-reckon it by chaining pieces together, so position cannot be used, but there is another way. By using the slope, bank, and isdiagonal properties of the track before and the track after, some track elements can be saved in TD6 as long as the track they are aliasing to does not have the same start and end slope, bank, and diag properties. There is no need to alias to a similarly-designed piece, for example LeftSmallZeroGRollUp can alias to Flat and RightSmallZeroGRollUp can alias to StationStart, and because the track shape is unique, LeftSmallZeroGRollDown can also alias to Flat and RightSmallZeroGRollDown can also alias to StationStart.

There would be a map table to map track elements between each other. When serializing TD6, it would do a simple swap of the source value to the target value. When deserializing, when importing a track element it determines if there is a map of the track element, then it gets the RTD of the current track element and the track element two pieces previous (and the RTD of the first element and of the previous element, if it is the final element). It then checks the RTD of the source piece and checks that its start properties match the end propertied and its end properties match the start properties.


There are a few caveats with this method: ~~the first and last element cannot be a mappable track element. It throws things in all sorts of disarray if the first element can be mappable~~ which is what happens with reverse-incline shuttle mode coasters. The track elements chosen to alias to should be lesser-used elements to optimize how much searching the game does. 

There are also some requirements for implementation:
1. Tests must be created to ensure the feature works as intended for all new track elements added
2. The OpenRCT2 team must approve of this method, because it will break backwards-compatibility with RCT2 on rides that use new elements (expected) and it is somewhat of a hack.


Resolving the issue of reverse-incline shuttle mode coasters requires one bit in the TD6 format somewhere. This one bit will do one of two things:
1. Signal that the first element in the chain is not the start of the coaster. This element is an "anchor" element that is added to the start of the ride that allows the real first element to be interpreted correctly. After the real first element is interpreted, the anchor element is dropped from the list of elements.
2. Signal that two track element structs will be used to hold the data. The first holds the LSB of TrackElemType and the track's properties. The second holds the MSB of TrackElemType and track properties are garbage.

There is a flags field, but even so, this level of refactors is polishing a turd.
